### PR TITLE
 port Austin's buffer mapping tests

### PR DESCRIPTION
**Pull request by kainino0x**
... more or less. Same-ish but can't use setSubData, and parameterized the tests with the tools available in the CTS.

Deduplicated with the other buffer tests.

### CODE REVIEW

**Comment by kainino0x** [comment]
@austinEng ping for review when you get a chance

File: src/suites/cts/buffers/map_oom.spec.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+export const description = ``;
+
+import { TestGroup } from '../../../framework/index.js';
+import { GPUTest } from '../gpu_test.js';
+
+const bufferDesc: GPUBufferDescriptor = {
+  size: Number.MAX_SAFE_INTEGER,
+  usage: 2, // GPUBufferUsage.MAP_WRITE (TODO: not in scope during test crawling in node)
```
Comment (austinEng) [inline src/suites/cts/buffers/map_oom.spec.ts]: Would it work if we just made this a function `const getBufferDesc = () => ({ size, usage: MAP_WRITE })` ?

File: src/suites/cts/buffers/mapping_test.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import { GPUTest } from '../gpu_test.js';
+
+export class MappingTest extends GPUTest {
+  checkMapWrite(buffer: GPUBuffer, arrayBuffer: ArrayBuffer, size: number): Promise<void> {
```
Comment (austinEng) [inline src/suites/cts/buffers/mapping_test.ts]: nit: Looking at the function in isolation, it's hard to see that `arrayBuffer` is the mapped buffer. Maybe name it `mappedContents`?

File: src/suites/cts/buffers/mapping_test.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    this.checkMapWriteZeroed(arrayBuffer, size);
+
+    const view = new Uint32Array(arrayBuffer);
```
Comment (austinEng) [inline src/suites/cts/buffers/mapping_test.ts]: nit: `mappedView` ?

File: src/suites/cts/buffers/map.spec.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+g.test('mapWriteAsync', async t => {
+  const size = t.params.size;
+  const buffer = t.device.createBuffer({
+    size,
+    usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.MAP_WRITE,
+  });
+
+  const arrayBuffer = await buffer.mapWriteAsync();
+  t.checkMapWriteZeroed(arrayBuffer, size);
```
Comment (austinEng) [inline src/suites/cts/buffers/map.spec.ts]: Can this test use the `checkMapWrite` helper?

File: src/suites/cts/buffers/create_mapped.spec.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+g.test('createBufferMapped', async t => {
+  const size = t.params.size;
+  const [buffer, arrayBuffer] = t.device.createBufferMapped({
+    size,
+    usage: GPUBufferUsage.COPY_SRC | (t.params.mappable ? GPUBufferUsage.MAP_WRITE : 0),
+  });
+  await t.checkMapWrite(buffer, arrayBuffer, size);
+}).params(
+  pcombine([
+    poptions('size', [12, 512 * 1024]), //
```
Comment (austinEng) [inline src/suites/cts/buffers/create_mapped.spec.ts]: nit: Empty comment?

File: src/suites/cts/buffers/create_mapped.spec.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+g.test('createBufferMappedAsync', async t => {
+  const size = t.params.size;
+  const [buffer, arrayBuffer] = await t.device.createBufferMappedAsync({
+    size,
+    usage: GPUBufferUsage.COPY_SRC | (t.params.mappable ? GPUBufferUsage.MAP_WRITE : 0),
+  });
+  await t.checkMapWrite(buffer, arrayBuffer, size);
+}).params(
+  pcombine([
+    poptions('size', [12, 512 * 1024]), //
```
Comment (austinEng) [inline src/suites/cts/buffers/create_mapped.spec.ts]: nit: Empty comment?

File: src/suites/cts/buffers/map_oom.spec.ts (hunk 1 above)
Comment (kainino0x) [reply src/suites/cts/buffers/map_oom.spec.ts]: Good idea.

File: src/suites/cts/buffers/create_mapped.spec.ts (hunk 5 above)
Comment (kainino0x) [reply src/suites/cts/buffers/create_mapped.spec.ts]: It's here to keep the formatter from combining these lines.

**Comment by kainino0x** [comment]
Thanks!

**Pull request merged**